ddtrace/tracer/sampler.go

```diff
@@ -147,3 +154,249 @@ func (ps *prioritySampler) apply(spn *span) {
+			rate, err := v.Rate.Float64()
+			if err != nil {
+				log.Warn("error parsing rule: invalid rate: %v", err)
+				continue
+			}
+			switch {
+			case v.Service != "" && v.Operation != "":
+				rules = append(rules, ServiceOperationRule(v.Service, v.Operation, rate))
+			case v.Service != "":
+				rules = append(rules, ServiceRule(v.Service, rate))
+			case v.Operation != "":
+				rules = append(rules, OperationRule(v.Operation, rate))
+			}
+		}
+	}
+	validRules := make([]SamplingRule, 0, len(rules))
+	for _, v := range rules {
+		if !(v.Rate >= 0.0 && v.Rate <= 1.0) {
+			log.Warn("ignoring rule %+v: rate is out of range", v)
+			continue
+		}
+		validRules = append(validRules, v)
+	}
+	return validRules
+}
+
+// sampleRate returns the rate to apply when the rate sampler's rules didn't match.
+// A zero value means the rate sampler
+func sampleRate() float64 {
+	const defaultRate = 0.0
+	v := os.Getenv("DD_TRACE_SAMPLE_RATE")
+	if v == "" {
+		return defaultRate
+	}
+	r, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate %f because DD_TRACE_SAMPLE_RATE is invalid: %v", defaultRate, err)
+		return defaultRate
+	}
+	if r >= 0.0 && r <= 1.0 {
+		return r
+	}
+	log.Warn("using default rate %f because provided value is out of range: %f", defaultRate, r)
+	return defaultRate
+}
+
+// newRateLimiter configures and returns a *rate.Limiter instance that is used when applying sampling rules.
+// The limit can be set with the DD_TRACE_RATE_LIMIT environment variable. Invalid values are ignored with
+// a warning message.
+func newRateLimiter(r float64) *rate.Limiter {
+	defaultLimiter := rate.NewLimiter(rate.Inf, 0)
+	v := os.Getenv("DD_TRACE_RATE_LIMIT")
+	if v == "" {
+		return defaultLimiter
+	}
+	l, err := strconv.ParseFloat(v, 64)
+	if err != nil {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is invalid: %v", err)
+		return defaultLimiter
+	}
+	if l < 0.0 {
+		log.Warn("using default rate limit because DD_TRACE_RATE_LIMIT is negative: %f", l)
+		return defaultLimiter
+	}
+	return rate.NewLimiter(rate.Limit(l), int(math.Ceil(r*l)))
+}
+
+// apply uses the sampling rules to determine the sampling rate for the
+// provided span. If the rules don't match, and a default rate hasn't been
+// set using DD_TRACE_SAMPLE_RATE, then it returns false and the span is not
+// modified.
+func (rs *rulesSampler) apply(span *span) bool {
+	var matched bool
+	rate := rs.rate
+	for _, rule := range rs.rules {
+		if rule.match(span) {
+			matched = true
+			rate = rule.Rate
+			break
+		}
+	}
+	if !matched && rate == 0.0 {
+		// no matching rule or global rate, so we want to fall back
+		// to priority sampling
+		return false
+	}
+	// rate sample
+	span.SetTag("_dd.rule_psr", rate)
+	if !sampledByRate(span.TraceID, rate) {
+		span.SetTag(ext.SamplingPriority, ext.PriorityAutoReject)
+		return true
+	}
+	// global rate limit and effective rate calculations
+	rs.mu.Lock()
+	defer rs.mu.Unlock()
+	ts := time.Now()
```

Comment:
This is a fairly expensive operation. Can we make `apply` take `time.Time` as an argument and pull it from [here](https://github.com/DataDog/dd-trace-go/blob/f3ef8903208cf44abcfaf44dcbc1d6cef198235d/ddtrace/tracer/tracer.go#L225) ? It will also help you test this bit.

Reply:
I've changed it to take a timestamp, but haven't refactored it all the way up to the `now()` function.

For the common case, when `tracer.StartTime` isn't used, we can reduce it to one time.Now() call. When `tracer.StartTime` is used, we can't use that timestamp for the rate-limiting decision, so it needs an updated value.
Do you think that's worth changing?

Reply:
I think it's fine to not spend the time at this point to change it. We can look at it again later once this actually starts to get used more heavily and potentially has an impact on performance.